penguin/engine.py

```
@@ -86,6 +86,123 @@
+    """Return an empty truncation payload matching API telemetry shape."""
+
+    return {
+        "total_truncations": 0,
+        "messages_removed": 0,
+        "tokens_freed": 0,
+        "by_category": {},
+        "recent_events": [],
+    }
+
+
+def _truncations_from_tracker(tracker: Any) -> Dict[str, Any]:
```

Ruff (0.15.13)

[warning] Dynamically typed expressions (typing.Any) are disallowed in `tracker`

(ANN401)